scipy/sparse/_compressed.py

```diff
@@ -500,13 +500,31 @@ def _matmul_multivector(self, other):
         fn(M, N, n_vecs, self.indptr, self.indices, self.data,
            other.ravel(), result.ravel())
 
+        if self.ndim == 1:
+            return result.reshape((n_vecs,))
         return result
 
     def _matmul_sparse(self, other):
-        M, K1 = self.shape
-        K2, N = other.shape
+        M, K1 = self._shape_as_2d
+        # if other is 1d, treat as a **column**
+        o_ndim = other.ndim
+        if o_ndim == 1:
+            # convert 1d array to a 2d column when on the right of @
+            other = other.reshape((1, other.shape[0])).T  # Note: converts to CSC
+        K2, N = other._shape
+
+        # find new_shape
+        if self.ndim == 2:
+            if o_ndim == 2:
+                new_shape = (M, N)
+            else:
+                new_shape = (M,)
+        elif o_ndim == 2:
+            new_shape = (N,)
+        else:
+            new_shape = ()
```

Comment:
```
new_shape = ()
if self.ndim == 2:
    new_shape += (M,)
if o_ndim == 2:
    new_shape += (N,)
```

Reply:
We could instead use `new_shape = tuple(MN for (MN, nd) in [(M, self.ndim), (N, o_ndim)] if nd == 2)`,
but I think the if statements are easier to figure out what is going on. So that's what I pushed up.